packages/file_selector/file_selector_android/android/src/main/java/dev/flutter/packages/file_selector_android/FileSelectorApiImpl.java

```diff
@@ -0,0 +1,343 @@
+      @NonNull List<String> extensions,
+      @NonNull
+          GeneratedFileSelectorApi.Result<List<GeneratedFileSelectorApi.FileResponse>> result) {
+    final Intent intent = testProxy.newIntent(Intent.ACTION_OPEN_DOCUMENT);
+    intent.addCategory(Intent.CATEGORY_OPENABLE);
+    intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true);
+
+    setMimeTypes(intent, mimeTypes, extensions);
+
+    try {
+      if (initialDirectory != null) {
+        trySetInitialDirectory(intent, initialDirectory);
+      }
+      tryStartActivityForResult(
+          intent,
+          OPEN_FILES,
+          new OnResultListener() {
+            @Override
+            public void onResult(int resultCode, @Nullable Intent data) {
+              if (resultCode == Activity.RESULT_OK && data != null) {
+                // Only one file was returned.
+                final Uri uri = data.getData();
+                if (uri != null) {
+                  final GeneratedFileSelectorApi.FileResponse file = toFileResponse(uri);
+                  if (file != null) {
+                    result.success(Collections.singletonList(file));
+                  } else {
+                    result.error(new Exception("Failed to read file: " + uri));
+                  }
+                }
+
+                // Multiple files were returned.
+                final ClipData clipData = data.getClipData();
+                if (clipData != null) {
+                  final List<GeneratedFileSelectorApi.FileResponse> files =
+                      new ArrayList<>(clipData.getItemCount());
+                  for (int i = 0; i < clipData.getItemCount(); i++) {
+                    final ClipData.Item clipItem = clipData.getItemAt(i);
+                    final GeneratedFileSelectorApi.FileResponse file =
+                        toFileResponse(clipItem.getUri());
+                    if (file != null) {
+                      files.add(file);
+                    } else {
+                      result.error(new Exception("Failed to read file: " + uri));
+                      return;
+                    }
+                  }
+                  result.success(files);
+                }
+              } else {
+                result.success(new ArrayList<>());
+              }
+            }
+          });
+    } catch (Exception exception) {
+      result.error(exception);
+    }
+  }
+
+  @Override
+  public void getDirectoryPath(
+      @Nullable String initialDirectory, @NonNull GeneratedFileSelectorApi.Result<String> result) {
+    if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
+      throw new UnsupportedOperationException(
+          "Selecting a directory is only supported on versions >= android.os.Build.VERSION_CODES.LOLLIPOP");
+    }
+
+    final Intent intent = testProxy.newIntent(Intent.ACTION_OPEN_DOCUMENT_TREE);
+    try {
+      if (initialDirectory != null) {
+        trySetInitialDirectory(intent, initialDirectory);
+      }
```

Comment:
Rather than repeat this check in every method, we could make the helper take a nullable directory, and no-op in that case.